python/rmm/_lib/device_buffer.pyx

```diff
@@ -78,9 +71,6 @@ cdef class DeviceBuffer:
             else:
                 self.c_obj.reset(new device_buffer(c_ptr, size, stream.view()))
```

Comment:
Would we need to possibly synchronize here in the case that `c_ptr` is a host ptr? Or are we saying that when working with ptrs that we do everything asynchronously?

Reply:
That's a good question, I don't know what's the correct answer from the RMM perspective. I think from the Python perspective we should indeed synchronize, but for that I think we would need `cudaPointerGetAttributes` which isn't exposed to Python as far as I see. Should I expose and use `cudaPointerGetAttributes` or is there a simpler way I'm overlooking?

Reply:
I think that's unfortunately the best bet for now

Reply:
Or we can pessimistically always synchronize if constructing from a pointer in Python for now. I'm not sure if this API gets actually used in practice.

Reply:
It gets used in Distributed: https://github.com/dask/distributed/blob/3d53801f1ba7b2d9b59804548162efc3bc57f857/distributed/protocol/rmm.py#L46

Anyway, it's probably best to try and play safe and efficient, so I'm gonna attempt the `cudaPointerGetAttributed` path, if we have a more appropriate way we can always update that later.

Reply:
I'm thinking we should remove pointers from the top level API and just handle `__cuda_array_interface__` vs `__array_interface__` and `memoryview` objects which can then also guide our synchronization behavior as well.

Reply:
Agreed. Though it's worth noting we do rely on these in a few places in Dask and cuDF. So we should discuss how we deprecate them, but yeah that seems like a good path forward.

Reply:
Argh, I just realized `cudaPointerGetAttributes` will not suffice here:

> If pointer was not allocated in, mapped by or registered with context supporting unified addressing cudaErrorInvalidValue is returned.

A hacky solution would be to just assume that if `cudaErrorInvalidValue` is raised to then assume it's a host pointer and we synchronize. Thoughts on this, or other ideas perhaps?

Reply:
I think we should just always synchronize when constructing from a pointer for now.

We'll likely rework this in the future to get rid of pointer handling directly in Python anyway.

Reply:
Why do you need to synchronize when constructing from a host pointer? Host memory is not stream ordered, so you don't have to wait on it. And all methods on the `DeviceBuffer` are stream ordered, so any accesses from the device or all copies will be stream ordered. Synchronizing here means you lose the asynchronous benefits of stream ordering.  The only place you really need to synchronize a stream-ordered entity is when returning a scalar value to the host thread after copying it DeviceToHost.

Reply:
If we don't synchronize here the host pointer could be mutated or freed before the copy finishes, no?

Reply:
The user was smart enough to pass a stream, meaning they want this to be stream ordered. They should be smart enough not to modify or delete the thing they are copying without synchronizing first.

The whole point of stream ordering is so that users who know what they are doing can achieve stall-free pipelines on the GPU. If we synchronize their streams everywhere, we kill that capability.

Reply:
This is the path of construction if a pointer is passed regardless of whether the user passed a stream or not, hence why I think we need to synchronize in the case of the default stream.

Reply:
As per the conversation @kkraus14 and I had offline, I've added the synchronization back but only when a pointer is passed, I adjusted the note accordingly. For now, we assume any pointer needs synchronization, regardless of being host or device pointer. As Keith noted too, this is most often used by host copies, as generally we would rely on `__cuda_array_interface__` for D2D.